### PR TITLE
Match similar trees

### PR DESCRIPTION
**Pull request by pubrrr**

### CODE REVIEW

File: src/domain/mod.rs [hunk 1]
```diff
@@ -1,5 +1,5 @@
 use std::collections::HashMap;
-use std::ops::{Add, Div, Mul, Neg, Sub};
+use std::ops::{Add, Deref, Div, Mul, Neg, Sub};
```
Comment (ekinimo) [inline src/domain/mod.rs:2]: why Deref? what does it do ?

File: src/domain/mod.rs [hunk 2]
```diff
@@ -270,34 +276,34 @@ impl Leaf {
...
-pub struct BinaryNode<'a> {
-    left: &'a Id,
-    right: &'a Id,
+pub struct BinaryNode {
+    left: Id,
+    right: Id,
     operator: BinaryOperator,
     root: Id,
-    syntax_tree: &'a AbstractSyntaxTree,
+    syntax_tree: Box<AbstractSyntaxTree>,
```
Comment (ekinimo) [inline src/domain/mod.rs:297]: rather use Rc here?

File: src/domain/mod.rs (hunk 1 above)
Comment (pubrrr) [reply src/domain/mod.rs:2]: It gets a &T out of a Box<T>

File: src/domain/mod.rs (hunk 2 above)
Comment (pubrrr) [reply src/domain/mod.rs:297]: Why exactly should cloning RC be cheaper than cloning Box?

**Pull request merged**